Support content->slot conversions of arbitrary cooperating elements.

rictic opened the pull request:
<!--
  Thanks for the PR!

  If this change has a user visible change (including
  bug fixes, new features, etc) please describe the change in
  CHANGELOG.md.

  If the change is an entirely package-internal reshuffling/refactoring
  should the change not be described in the CHANGELOG.

  Consider also updating the README.

  More info: http://keepachangelog.com/en/0.3.0/
 -->

 - [x] CHANGELOG.md has been updated

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/polymer/polymer-linter/112)
<!-- Reviewable:end -->

valdrinkoshi commented on test/content-to-slot-usages/after-fixes.html:
```diff
@@ -61,3 +61,27 @@
+
+<dom-module id="foo-bar">
+  <template>
+    <slot></slot>
```
nit: this should be moved as last slot as it has no name (would distribute all content in here)

valdrinkoshi commented on test/content-to-slot-usages/before-fixes.html:
```diff
@@ -61,3 +61,27 @@
+
+<dom-module id="foo-bar">
+  <template>
+    <slot></slot>
```
nit: this should be moved below as last slot as it has no name (would distribute all content in here)

valdrinkoshi commented on src/polymer/content-to-slot-usages.ts:
```diff
@@ -84,6 +84,35 @@ addPredicate(
     'paper-dropdown-menu',
     [{selector: '.dropdown-content', slot: 'dropdown-content'}]);
 
+function getContentDescriptors(tagName: string, document: Document):
```
documentation needed 📄

valdrinkoshi commented on src/polymer/content-to-slot-usages.ts:
```diff
@@ -84,6 +84,35 @@ addPredicate(
+    imported: true,
+    externalPackages: true
+  });
+  if (element && element.domModule) {
```
what about elements in the main document? should that be a separate run?

rictic commented:
I added some more documentation, and added warnings for unsupported CSS expressions in `old-content-selector`, to handle the case pointed out here: https://github.com/Polymer/polymer-linter/pull/109#discussion_r147539419

The heavier processing of selectors made it seem worth it to cache results, so I added a weak map cache based on the Element object (which is immutable, so whenever it changes we'll get a different instance and thus a different entry in the weak map).

PTAL!

valdrinkoshi commented on src/polymer/content-to-slot-usages.ts:
```diff
@@ -172,27 +121,219 @@ class ContentToSlot extends HtmlRule {
-    return p.hasAttr(simpleSelector.slice(1, -1));
+}
+
+type Result<Good, Bad> = {
```
..and the Ugly? 😁

rictic replied:
:P

actually… when you've got a good result, a recoverable error, and an unrecoverable error… not a bad idea

valdrinkoshi commented on CHANGELOG.md:
```diff
@@ -7,8 +7,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## Unreleased
 - Support warning about and automatically fixing some simple uses of some 1.0 elements to their 2.0 usage styles.
+  - To add this fixable lint warning for your own elements, simply add the `old-content-selector` attribute to your new `<slot>` element. Set its value to the value that your old `<content>` element had for its `select` attribute.
```
I'd add an example here, wdyt of something like
```html
<!-- <content select=".some-content"></content> -->
<slot old-content-selector=".some-content"></slot>

rictic replied:
+1, wrote it out more explicitly

valdrinkoshi commented on src/polymer/content-to-slot-usages.ts:
```diff
@@ -172,27 +121,219 @@ class ContentToSlot extends HtmlRule {
+                         })));
 }
 
+addPredicate(
```
Probably material for a separate PR, but I'd define these predicates externally (e.g. a json file) and make the path configurable/extensible. This would allow element authors to share their upgrades to users of their elements.

rictic replied:
I'd prefer to instead make changes to our elements to add the `old-content-selector` attribute and to remove this hard coded set. I added it in while I was initially hacking on this stuff. `old-content-selector` is better in every way.

I'll do that in a separate PR.
